sphinx/testing/fixtures.py

```diff
@@ -2,216 +2,470 @@
 
 from __future__ import annotations
 
+import os
 import shutil
 import subprocess
 import sys
-from collections import namedtuple
-from io import StringIO
-from typing import TYPE_CHECKING, Optional
+import warnings
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Optional, cast
 
 import pytest
 
-from sphinx.testing.util import SphinxTestApp, SphinxTestAppWrapperForSkipBuilding
+from sphinx.deprecation import RemovedInSphinx90Warning
+from sphinx.testing.internal.cache import AppInfo, LegacyModuleCache, ModuleCache
+from sphinx.testing.internal.isolation import Isolation
+from sphinx.testing.internal.markers import (
+    AppLegacyParams,
+    AppParams,
+    process_isolate,
+    process_sphinx,
+    process_test_params,
+)
+from sphinx.testing.internal.pytest_util import (
+    TestRootFinder,
+    find_context,
+    get_mark_parameters,
+)
+from sphinx.testing.internal.pytest_xdist import is_pytest_xdist_enabled
+from sphinx.testing.util import (
+    SphinxTestApp,
+    SphinxTestAppLazyBuild,
+    SphinxTestAppWrapperForSkipBuilding,
+    strip_escseq,
+)
 
 if TYPE_CHECKING:
-    from collections.abc import Callable, Generator
+    from collections.abc import Generator
+    from io import StringIO
     from pathlib import Path
-    from typing import Any
+    from typing import Any, Final, Union
 
-DEFAULT_ENABLED_MARKERS = [
+    from _pytest.nodes import Node as PytestNode
+
+    from sphinx.testing.internal.isolation import IsolationPolicy
+    from sphinx.testing.internal.markers import TestParams
+
+    AnySphinxTestApp = Union[SphinxTestApp, SphinxTestAppWrapperForSkipBuilding]
+    AnyAppParams = Union[AppParams, AppLegacyParams]
+
+DEFAULT_ENABLED_MARKERS: Final[list[str]] = [
     (
-        'sphinx(builder, testroot=None, freshenv=False, confoverrides=None, tags=None, '
```

Comment:
Note that the current implementation (on master) assumes that 'testroot' is a keyword argument (it nevers checks for the 1st positional argument).